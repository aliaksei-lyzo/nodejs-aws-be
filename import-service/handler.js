module.exports = {
  importProductsFile: async function(event) {
    console.log('EVENT', event);
    try {
      const AWS = require('aws-sdk');
      const BUCKET = 'speedworks-products';

      const filename = event.queryStringParameters.name;
      console.log('filename', filename)
      if (!filename) throw new Error('Missing required parameter');
      const s3 = new AWS.S3({ region: 'eu-west-1' });
      const catalogPath = `uploaded/${filename}`;
      
      const params = {
        Bucket: BUCKET,
        Key: catalogPath,
        Expires: 60,
        ContentType: 'text/csv',
      };
      
      const signedUrl = await s3.getSignedUrlPromise('putObject', params);

      return {
        statusCode: 200,
        headers: { 'Access-Control-Allow-Origin': '*' },
        body: signedUrl,
      };

    } catch (error) {
      console.log('ERROR', error)
      return {
        statusCode: 400,
        headers: { 'Access-Control-Allow-Origin': '*' },
        body: JSON.stringify({
          success: false,
          error: JSON.stringify(error, Object.getOwnPropertyNames(error)),
        })
      }
    }
  },

  importFileParser: async function(event) {
    console.log('EVENT', JSON.stringify(event));
    try {
      const AWS = require('aws-sdk');
      const csv = require('csv-parser');
      const BUCKET = 'speedworks-products';

      const s3 = new AWS.S3({ region: 'eu-west-1' });
      const record = event.Records[0];

      const products = [];
      console.log(`record`, record);
      const csvStream = s3.getObject({ Bucket: BUCKET, Key: record.s3.object.key })
      .createReadStream();
      console.log('csvStream', JSON.stringify(csvStream));
      await new Promise((res) => {
        csvStream
        .on('error', (err) => {
          console.log('Error in stream');
          console.log(JSON.stringify(err));
         })
        .pipe(csv())
        .on('data', product => {
          products.push(product);
        })
        .on('end', async () => {
          console.log('Parsed successfully');
          console.log(products);
          res('okay');
        })
      });

      console.log('Moving file to parsed');
      await s3.copyObject({
        Bucket: BUCKET,
        CopySource: `/${BUCKET}/${record.s3.object.key}`,
        Key: record.s3.object.key.replace('uploaded', 'parsed')
      }).promise();

      console.log('Deleting file from uploaded');
      await s3.deleteObject({
        Bucket: BUCKET,
        Key: record.s3.object.key,
      }).promise();

      console.log('All done!');
      return 200
    } catch (e) {
      console.log('ERROR', JSON.stringify(e, Object.getOwnPropertyNames(e)));
      return 500;
    }
    
  }

}