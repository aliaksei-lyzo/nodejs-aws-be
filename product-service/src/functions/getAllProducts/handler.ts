import 'source-map-support/register';

import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/apiGateway';
import { successfulResponse, errorResponse } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import logger from '@libs/logger';

import schema from './schema';

import { getDbClient } from '@libs/pgClient';
import { queryAllProducts } from '../../services/product.service';

export const getAllProducts: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event) => {
  logger.info(`Receiving event: ${JSON.stringify(event)}`);
  try {
    const db = getDbClient();
    const productsList = await queryAllProducts(db);
    return successfulResponse({
      products: productsList,
      event,
    });
  } catch(e) {
    return errorResponse(500, e);
  }
}

export const main = middyfy(getAllProducts);
