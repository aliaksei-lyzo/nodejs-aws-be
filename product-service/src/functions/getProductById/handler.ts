import 'source-map-support/register';

import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/apiGateway';
import { successfulResponse, errorResponse } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import logger from '@libs/logger';

import schema from './schema';

import { productsListMock, Product } from '../../libs/productsMock';

const getProductById: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event) => {
  logger.info(`Receiving event: ${JSON.stringify(event)}`);
  try {
    const productId = event?.pathParameters?.productId;

    if (!productId) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          success: false,
          message: 'Bad request - required parameter missing',
        }),
      };
    }

    // imitate DB request. Later this will be extracted to a separate service
    const productsList: Product[] = await new Promise((res) => {
      setTimeout(() => {
        res(productsListMock);
      }, 1000);
    })

    const product = productsList.find(({ id }) => id === productId);
    if (product) {
      return successfulResponse({
        product,
        event,
      });
    }

    return errorResponse(404, new Error(`Product with id ${productId} not found`));

  } catch(e) {
      return errorResponse(500, e);
  }
}

export const main = middyfy(getProductById);
