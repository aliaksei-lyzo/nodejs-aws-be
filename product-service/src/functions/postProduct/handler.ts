import 'source-map-support/register';

import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/apiGateway';
import { successfulResponse, errorResponse } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import logger from '@libs/logger';

import schema from './schema';

import { getDbClient } from '@libs/pgClient';
import { queryAddProduct } from '../../services/product.service';

import { Product } from '../../types';

export const postProduct: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event) => {
  logger.info(`Receiving event: ${JSON.stringify(event)}`);
  const product = event.body.product;
  const number = event.body.number || 0;
  if (!product || !product.title || !product.price) return errorResponse(400, new Error('missing required product parameters'));
  try {
    const db = getDbClient();
    const newProduct: Product = await queryAddProduct(db, product, number);
    return successfulResponse({
      product: newProduct,
      event,
    });
  } catch(e) {
    return errorResponse(500, e);
  }
}

export const main = middyfy(postProduct);
