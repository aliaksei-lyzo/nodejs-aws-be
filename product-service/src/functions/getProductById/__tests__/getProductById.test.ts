import { getProductById } from '../handler';
import * as utils from 'aws-lambda-test-utils';

describe('Test getAllProducts lambda', () => {
  it('should return products list', async () => {
    const event = utils.mockEventCreator.createAPIGatewayEvent({ pathParameters: { productId: '7567ec4b-b10c-48c5-9345-fc73c48a80aa' } });
    const context = utils.mockContextCreator({}, () => {});
    expect(await getProductById(event, context, () => {})).toMatchSnapshot();
  });
});
