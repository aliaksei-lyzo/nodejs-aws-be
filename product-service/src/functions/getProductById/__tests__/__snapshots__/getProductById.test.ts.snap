// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test getAllProducts lambda should return products list 1`] = `
Object {
  "body": "{\\"success\\":true,\\"response\\":{\\"product\\":{\\"count\\":4,\\"description\\":\\"Short Product Description1\\",\\"id\\":\\"7567ec4b-b10c-48c5-9345-fc73c48a80aa\\",\\"price\\":2.4,\\"title\\":\\"ProductOne\\"},\\"event\\":{\\"resource\\":\\"/assets\\",\\"path\\":\\"default/path\\",\\"httpMethod\\":\\"GET\\",\\"headers\\":{\\"default-header\\":\\"default\\"},\\"queryStringParameters\\":{\\"query\\":\\"default\\"},\\"pathParameters\\":{\\"productId\\":\\"7567ec4b-b10c-48c5-9345-fc73c48a80aa\\"},\\"stageVariables\\":{\\"ENV\\":\\"test\\"},\\"requestContext\\":{\\"accountId\\":\\"1234\\",\\"resourceId\\":\\"snmm5d\\",\\"stage\\":\\"test-invoke-stage\\",\\"requestId\\":\\"test-invoke-request\\",\\"identity\\":{\\"cognitoIdentityPoolId\\":null,\\"accountId\\":\\"1234\\",\\"cognitoIdentityId\\":null,\\"caller\\":\\"1234\\",\\"apiKey\\":\\"test-invoke-api-key\\",\\"sourceIp\\":\\"test-invoke-source-ip\\",\\"accessKey\\":\\"1234\\",\\"cognitoAuthenticationType\\":null,\\"cognitoAuthenticationProvider\\":null,\\"userArn\\":\\"arn:aws:iam::1234:user/test_user\\",\\"userAgent\\":\\"Apache-HttpClient/4.5.x (Java/1.8.0)\\",\\"user\\":\\"1234\\"},\\"resourcePath\\":\\"default/path\\",\\"httpMethod\\":\\"GET\\",\\"apiId\\":\\"1234\\"},\\"body\\":\\"default body\\",\\"isBase64Encoded\\":false}}}",
  "statusCode": 200,
}
`;
